File: 백준/레벨Gold V/1092. 배/배.py
from sys import stdin
import math

input = stdin.readline

# 입력 처리
N = int(input().rstrip())
cranes = list(map(int, input().split()))
M = int(input().rstrip())
boxes = list(map(int, input().split()))

# 내림차순 정렬
cranes.sort(reverse=True)
boxes.sort(reverse=True)

if boxes[0] > cranes[0]:
    print(-1)
else:
    # 최소 시간 : 모든 박스 / 모든 크레인
    max_time = math.ceil(M / N)
    b_idx = 0

    for c_idx in range(1, N):
        # 현재 crane이 들 수 있는 가장 큰 box를 확인
        while b_idx < M and cranes[c_idx] < boxes[b_idx]:
            b_idx += 1

        # 해당 박스 범위 / 해당 크레인 범위
        cur_time = math.ceil(b_idx / c_idx)
        max_time = max(max_time, cur_time)

        if b_idx == M:
            break
    
    print(max_time)

File: 백준/레벨Gold V/1092. 배/배.ts
// @ts-ignore
import fs = require('fs');

const lines = fs.readFileSync('/dev/stdin').toString().split('\n');

// 입력 처리
const N: number = parseInt(lines[0].trim());
const cranes: number[] = lines[1].trim().split(' ').map(Number);
const M: number = parseInt(lines[2].trim());
const boxes: number[] = lines[3].trim().split(' ').map(Number);

// 내림차순 정렬
cranes.sort((a, b) => b - a);
boxes.sort((a, b) => b - a);

if (boxes[0] > cranes[0]) {
  console.log(-1);
} else {
  // 최소 시간 : 모든 박스 / 모든 크레인
  let maxTime: number = Math.ceil(M / N);
  let bIdx: number = 0;

  for (let cIdx = 1; cIdx < N; cIdx++) {
    // 현재 crane이 들 수 있는 가장 큰 box를 확인
    while (bIdx < M && cranes[cIdx] < boxes[bIdx]) {
      bIdx++;
    }

    // 현재 시간 : 박스 수 / 크레인 수
    const curTime: number = Math.ceil(bIdx / cIdx);
    maxTime = Math.max(maxTime, curTime);

    if (bIdx === M) break;
  }

  console.log(maxTime);
}